From 72ad12d1dc0601c05e615c73e280a993d6b367a4 Mon Sep 17 00:00:00 2001
From: Giuseppe Lettieri <g.lettieri@iet.unipi.it>
Date: Tue, 9 Apr 2019 17:24:55 +0200
Subject: [PATCH 3/4] bm: don't cross 64KiB boundaries

---
 hw/ide/pci.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/hw/ide/pci.c b/hw/ide/pci.c
index fe1ceeb0cd..dc4b8a3812 100644
--- a/hw/ide/pci.c
+++ b/hw/ide/pci.c
@@ -100,7 +100,17 @@ static int32_t bmdma_prepare_buf(IDEDMA *dma, int32_t limit)
              * PRDs from the guest, but ignore them. */
             sg_len = MIN(limit - s->sg.size, bm->cur_prd_len);
             if (sg_len) {
-                qemu_sglist_add(&s->sg, bm->cur_prd_addr, sg_len);
+                /* wrap around if the transfer crosses a 64KiB boundary */
+                uint64_t next_boundary =
+                    (bm->cur_prd_addr + 0x10000) & ~0xFFFF;
+                int delta = next_boundary - bm->cur_prd_addr;
+                if (delta < sg_len) {
+                    qemu_sglist_add(&s->sg, bm->cur_prd_addr, delta);
+                    qemu_sglist_add(&s->sg, bm->cur_prd_addr & ~0xFFFF,
+                            sg_len - delta);
+                } else {
+                    qemu_sglist_add(&s->sg, bm->cur_prd_addr, sg_len);
+                }
             }
 
             bm->cur_prd_addr += l;
-- 
2.17.1

